pkg/adapter/apiserver/adapter.go

```diff
@@ -0,0 +1,129 @@
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	eventsclient "github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing/pkg/reconciler"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"github.com/knative/pkg/controller"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// ReconcilerName is the name of the reconciler
+	ReconcilerName = "ApiServerSource"
+
+	controllerAgentName = "apiserver-source-adapter-controller"
+	updateEventType     = "dev.knative.apiserver.object.update"
+	deleteEventType     = "dev.knative.apiserver.object.delete"
+)
+
+// NewController initializes the controller and is called by the generated code
+// Registers event handlers to enqueue events
+func NewController(
+	opt reconciler.Options,
+	informer cache.SharedInformer,
+	lister cache.GenericLister,
+	eventsclient eventsclient.Client,
+	controlled bool) *controller.Impl {
+
+	r := &Reconciler{
+		Base:         reconciler.NewBase(opt, controllerAgentName),
+		lister:       lister,
+		eventsClient: eventsclient,
+	}
+	impl := controller.NewImpl(r, r.Logger, ReconcilerName, reconciler.MustNewStatsReporter(ReconcilerName, r.Logger))
+
+	r.Logger.Info("Setting up event handlers")
+
+	if controlled {
+		informer.AddEventHandler(reconciler.Handler(impl.EnqueueControllerOf))
+	} else {
+		informer.AddEventHandler(reconciler.Handler(impl.Enqueue))
+	}
+	return impl
+}
+
+// Reconciler reconciles an ApiServerSource object
+type Reconciler struct {
+	*reconciler.Base
+
+	eventsClient eventsclient.Client
+	lister       cache.GenericLister
+}
+
+// Reconcile sends a cloud event corresponding to the given key
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		r.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the resource with this namespace/name
+	original, err := r.lister.ByNamespace(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		r.Logger.Error("resource key in work queue no longer exists", zap.Any("key", key))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	object := original.(*duckv1alpha1.AddressableType)
+
+	eventType := updateEventType
+	timestamp := object.GetCreationTimestamp()
+	if object.GetDeletionTimestamp() != nil {
+		eventType = deleteEventType
+		timestamp = *object.GetDeletionTimestamp()
+	}
+
+	objectRef := corev1.ObjectReference{
+		APIVersion: object.APIVersion,
+		Kind:       object.Kind,
+		Name:       object.GetName(),
+		Namespace:  object.GetNamespace(),
+	}
+
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			ID:     string(object.GetUID()),
```

Comment:
`object.GetUID()` will always return the same value so the cloudevent will not be unique.

Reply:
it satisfies https://github.com/cloudevents/spec/blob/v0.2/spec.md#id constraints, in particular `MUST be unique within the scope of the producer`. @n3wscott what is your concern?